### PR TITLE
Max steps

### LINKED ISSUE #27: [Feature Request]: Specify Max Steps in ReAct and Reflexion
### Feature Description

In the prompts used in ReflexionCoT and ReflexionReAct and ReAct, we should specify the max number of steps. Within the prompt there should be a line that says: `You have a maximum of {n} steps.`

### Reason

The agent is unaware that the current setup has to finish in `n` steps. This specification in the prompt will help it at least come up with an answer by the end of the task generation/trajectory.

### PR DESCRIPTION
**Pull request by JoePehlke**
### 🤔 Reasoning

Allow for max_steps to be specified in the prompt for the agent. 

### 🚧 Changes

Functional changes - pass max_steps in both _build_agent_prompt and _prompt_agent. Add max_steps in format Prompts changes - Add "You have a maximum of {n} steps" agent changes - Pass max_tokens in _prompt_agent for thought and action. Only for react so far and no tests, just a print statement in functional.

### ✅ PR Checklist
Closes #27 
- [x] Using this PR template?
- [ x] Linked issue?
- [ x] Added feature?
    - [ ] Added/updated docs?
    - [ ] Added/updated tests?

### CODE REVIEW

**Comment by JoePehlke** [comment]
Not ready as of 2/23 6:22pm, needs testing and reflexion. First commits.

File: discussion_agents/cog/prompts/react.py [hunk 1]
```diff
@@ -1,76 +1,76 @@
...
-Action 3: Finish[yes]
-"""
+"""ReAct prompts and fewshot examples."""
+
+
+REACT_INSTRUCTION = """Solve a question answering task with interleaving Thought, Action, Observation steps. Thought can reason about the current situation, and Action can be three types: 
+(1) Search[entity], which searches the exact entity on Wikipedia and returns the first paragraph if it exists. If not, it will return some similar entities to search.
+(2) Lookup[keyword], which returns the next sentence containing keyword in the last passage successfully found by Search.
+(3) Finish[answer], which returns the answer and finishes the task.
+You have a maximum of {n} steps.
```
Comment (alckasoc) [inline discussion_agents/cog/prompts/react.py:8]:
```suggestion
You have a maximum of {max_steps} steps.

```

File: discussion_agents/cog/functional/react.py [hunk 2]
```diff
@@ -1,94 +1,99 @@
...
+        str: A formatted prompt template ready for use.
+    """
+    prompt = PromptTemplate.from_template(REACT_INSTRUCTION).format(
+        examples=REACT_WEBTHINK_SIMPLE6_FEWSHOT_EXAMPLES,
+        question=question,
+        scratchpad=scratchpad,
+        n=max_steps
+    )
+    #Verify max steps in prompt
+    #print(prompt)
```
Comment (alckasoc) [inline discussion_agents/cog/functional/react.py:34]: delete once done

File: discussion_agents/cog/functional/react.py [hunk 3]
```diff
@@ -1,94 +1,99 @@
...
+def _build_agent_prompt(question: str, scratchpad: str, max_steps: int) -> str:
+    """Constructs a prompt template for the agent.
+
+    This function formats a predefined prompt template (REACT_INSTRUCTION) with examples,
+    the provided question, and a scratchpad.
+
+    Args:
+        question (str): The question to be included in the prompt.
+        scratchpad (str): Additional scratchpad information to be included.
+        max_steps (int): Maximum steps allowed to process the question
```
Comment (alckasoc) [inline discussion_agents/cog/functional/react.py:23]:
```suggestion
        max_steps (int): Maximum steps allowed to process the question.
```

File: discussion_agents/cog/functional/react.py [hunk 4]
```diff
@@ -1,94 +1,99 @@
...
+        scratchpad (str): Additional scratchpad information to be included.
+        max_steps (int): Maximum steps allowed to process the question 
+    Returns:
+        str: A formatted prompt template ready for use.
+    """
+    prompt = PromptTemplate.from_template(REACT_INSTRUCTION).format(
+        examples=REACT_WEBTHINK_SIMPLE6_FEWSHOT_EXAMPLES,
+        question=question,
+        scratchpad=scratchpad,
+        n=max_steps
```
Comment (alckasoc) [inline discussion_agents/cog/functional/react.py:31]:
```suggestion
        max_steps=max_steps
```

File: discussion_agents/cog/functional/react.py [hunk 5]
```diff
@@ -1,94 +1,99 @@
...
+    """Generates a response from the LLM based on a given question and scratchpad.
+
+    This function creates a prompt using `_build_agent_prompt` and then gets the LLM's
+    output. The newline characters in the output are removed before returning.
+
+    Args:
+        llm (BaseChatModel): The language model to be prompted.
+        question (str): The question to ask the language model.
+        scratchpad (str): Additional context or information for the language model.
+        max_steps (int): Maximum steps allowed to process the question
```
Comment (alckasoc) [inline discussion_agents/cog/functional/react.py:48]:
```suggestion
        max_steps (int): Maximum steps allowed to process the question.
```

**Review by alckasoc — CHANGES_REQUESTED** [review]
Overall looks good. A couple syntax suggestions + make sure this all looks correct and is tested. I don't see any unit tests so far. Also, is there any reason why there are so many lines changed (~300+ ~300-)?

**Pull request closed without merging**